### PR TITLE
fpc: bump to 3.2.0

### PR DESCRIPTION
**Pull request by kencu**
bootstrap compiler is now accessed differently
via a dmg and pkg rather than a gzipped boostrap

previous patches no longer apply and appear
no longer needed

###### Type(s)

- [ ] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on

macOS 10.15.6 19G73
Xcode 11.6 11E708 

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL?

- [ ] checked your Portfile with `port lint`?
- [ ] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [ ] tested basic functionality of all binary files?

### CODE REVIEW

**Comment by kencu** [comment]
@kamischi -- are you around to take a look at this? Looks fairly neat and tidy, minimal changes.

Other than moving to the dmg bootstrap, the only minor change I made was I stripped out the hard-coding for 10.9 as we build this on older systems too.

**Comment by kencu** [comment]
This is a nomaintainer port, so I could just push this, but being polite to @kamischi given his extensive recent input, and perhaps @ryandesign wants a go at it too

File: lang/fpc/Portfile [hunk 1]
```diff
@@ -18,31 +18,41 @@ homepage            http://www.freepascal.org
...
 supported_archs     x86_64
 
 # doesn't install shared libraries
 installs_libs       no
 
 extract.only        ${src}
 
+# get the bootstrap compiler from the dmg
+post-extract {
+    system -W ${distpath} "hdiutil attach fpc-3.2.0.intel-macosx.dmg -noverify"
```
Comment (l2dy) [inline lang/fpc/Portfile:41]: Please specify a custom `-mountpoint`. The default mountpoint is not reliable.

https://github.com/macports/macports-base/blob/e82a6397e5a404ac157552129015b21557f57d53/src/port1.0/portextract.tcl#L119-L122

Comment (kencu) [reply lang/fpc/Portfile:41]: hey, thanks. Will do. By the way, have you seen any way to use MacPorts' built-in ```use_dmg``` functionality when you have another distfile to decompress as well? I tried several things but it appears I have to do it manually...

Comment (l2dy) [reply lang/fpc/Portfile]: MacPorts' built-in extract phase does not support multiple formats, so you have to do it manually.

BTW, `-readonly` and `-nobrowse` flags are also recommended, see the man page `hdiutil(1)` for more information.

File: lang/fpc/Portfile [hunk 2]
```diff
@@ -18,31 +18,43 @@ homepage            http://www.freepascal.org
...
+# get the bootstrap compiler from the dmg
+post-extract {
+    set dmg_mount [mkdtemp "/tmp/mports.XXXXXXXX"]
+    system -W ${distpath} "hdiutil attach fpc-3.2.0.intel-macosx.dmg -noverify -mountpoint ${dmg_mount}"
+    system -W ${workpath} "pkgutil --expand '${dmg_mount}/fpc-3.2.0-intel-macosx.pkg' ${workpath}/pkg"
+    system -W ${workpath}/pkg/fpc-3.2.0-intel-macosx.pkg "gzip -dc Payload | cpio -id"
+    ln -s ${workpath}/pkg/fpc-3.2.0-intel-macosx.pkg/usr/local/bin/ppcx64 ${workpath}/ppcx64
+    ln -s ${workpath}/pkg/fpc-3.2.0-intel-macosx.pkg/usr/local/bin/ppc386 ${workpath}/ppc386
+    system "hdiutil detach ${dmg_mount}"
+    system "rmdir ${dmg_mount}"
```
Comment (ryandesign) [inline lang/fpc/Portfile:48]: You can delete directories using `delete ...` rather than `system "rmdir ..."`.

File: lang/fpc/Portfile (hunk 1 above)
Comment (kencu) [reply lang/fpc/Portfile]: Yep, I can see why. Thanks. Guess. I should have just copied the whole enchilada out of the portextract.tcl file, but I was  distracted by the pkgutil business adding another layer :>

File: lang/fpc/Portfile (hunk 2 above)
Comment (kencu) [reply lang/fpc/Portfile:48]: Thanks, simpler.

**Comment by kamischi** [comment]
Hi.

I appreciate your work. For me, getting and extracting the disk image for the bootstrap compiler sounds a bit like overkill, although it offers both 32bit and 64 bit and will most likely provide it automatically in future versions. Personally, I prefer x86_64-macosx-10.9-ppcx64.tar.bz2, which is much smaller than the complete disk image and already available now. In case they are actually needed, 32bit and/or older systems could also be made available upstream. What is your idea about this?

I can confirm, that the patch files are no longer required.

My next step will be to add the Qt4pas package, which will require to re-add a fpc.cfg.patch, so that the compiler finds the Qt4pas framework, but time will tell. Can the option for the minimum version of OS X (10.9) -WM10.9 be added for systems >= 10.9?

Michael.

**Comment by kamischi** [comment]
Before I forget it:

Unfortunately, FreePascal has a Heisenbug (It's there as well as it's not there), when using parallel builds. It seems to depend on the general CPU load, but so far, it has not been possible to track down the issue. To be on the safe side, make all should be done as a single thread, like a `make -j 1`. How can this be achieved?

**Comment by kencu** [comment]
> Personally, I prefer x86_64-macosx-10.9-ppcx64.tar.bz2, which is much smaller than the complete disk image and already available now. 

Where is that, may I ask? I was unable to find such a thing, ergo the dmg shenanigans.



> In case they are actually needed, 32bit and/or older systems could also be made available upstream. What is your idea about this?

We like to support all the systems that can be supported, within reason.

> Can the option for the minimum version of OS X (10.9) -WM10.9 be added for systems >= 10.9?

It could -- but why, may I ask? It limits the build, and I am not quite sure why it is necessary to add it. I don't see any warnings or errors in this Catalina system without it, but I don't use fpc/lazarus as much as you obviously do.

**Comment by kamischi** [comment]
> Where is that, may I ask? I was unable to find such a thing, ergo the dmg shenanigans.

http://ftp.freepascal.org/dist/3.0.4/bootstrap/x86_64-macosx-10.9-ppcx64.tar.bz2

after some time it should be mirrored at sourceforge, but it seems to have been forgotten. Can you add the ftp.freepascal.org source?

Note, that using the 3.0.4 version for bootstrapping for 3.2.* is no mistake, but corresponds to the intended default scheme. 

> > In case they are actually needed, 32bit and/or older systems could also be made available upstream. What is your idea about this?
> 
> We like to support all the systems that can be supported, within reason.
> 
> > Can the option for the minimum version of OS X (10.9) -WM10.9 be added for systems >= 10.9?
> 
> It could -- but why, may I ask? It limits the build, and I am not quite sure why it is necessary to add it. I don't see any warnings or errors in this Catalina system without it, but I don't use fpc/lazarus as much as you obviously do.

I have added this some time ago, and haven't rechecked it. At its time, it was a simple harmless warning, but it made some people think, the compiler is non-functional. I'll have another look into this.

**Comment by kencu** [comment]
> Before I forget it:
> 
> Unfortunately, FreePascal has a Heisenbug (It's there as well as it's not there), when using parallel builds. It seems to depend on the general CPU load, but so far, it has not been possible to track down the issue. To be on the safe side, make all should be done as a single thread, like a `make -j 1`. How can this be achieved?

You do this
```
use_parallel_build no
```
which is already in the ```lazarus``` build I notice, but not in ```fpc```.

**Comment by kencu** [comment]
> > Where is that, may I ask? I was unable to find such a thing, ergo the dmg shenanigans.
> 
> http://ftp.freepascal.org/dist/3.0.4/bootstrap/x86_64-macosx-10.9-ppcx64.tar.bz2
> 
> after some time it should be mirrored at sourceforge, but it seems to have been forgotten. Can you add the ftp.freepascal.org source?

Hah -- I never would have gone digging around in there!

The ```ppcx64``` in the dmg works all the way back to 10.4, says the Readme. Is it actually different somehow, or is the README incorrect?

 Eventually, I plan to get this building back to 10.5 Intel and 10.4 PPC.

**Comment by kencu** [comment]
If found it here <ftp://ftp.freepascal.org/pub/fpc/dist/3.0.4/bootstrap/x86_64-macosx-10.9-ppcx64.tar.bz2>.

I'll change to use that, as it is for sure much smaller. Later on perhaps I'll sort out < 10.9.

**Pull request closed without merging**